### PR TITLE
convert a keyword search into a faceted term search

### PR DESCRIPTION
**Pull request by wluo**
- provided we can convert the search keyword into an authoritative facet term
- if more than one facet is matched, we'll use the first one with the highest priority configured

related to these pull request: https://github.com/GigaOM/gigaom/pull/5114 and https://github.com/misterbisson/scriblio-authority/pull/76

see https://github.com/GigaOM/legacy-pro/issues/3995

### CODE REVIEW

File: plugin/class-facets.php [hunk 1]
```diff
@@ -119,8 +118,27 @@ public function parse_query( $query )
 //print_r( $this->selected_facets );
 //echo "</pre>";
 
+		// detect if a keyword search could be converted to a facet search
+		if (
+			isset( scriblio()->options['redirect_keywords_to_tax_terms'] ) &&
```
Comment (misterbisson) [inline plugin/class-facets.php]: I know some of the terminology is mine, but we should use the `searchword` from the facet name. Perhaps `redirect_searchword_to_taxonomy`?

File: plugin/class-facets.php [hunk 2]
```diff
@@ -119,8 +118,27 @@ public function parse_query( $query )
...
+			scriblio()->options['redirect_keywords_to_tax_terms']
+		)
+		{
+			if ( list( $new_facet, $new_facet_obj ) = $this->can_redirect_keyword_search() )
+			{
+				// set up our selected facets to convert a keyword query
+				// (searchword facet) to a faceted term query for permalink()
+				unset( $this->selected_facets->searchword );
+
+				$this->selected_facets->$new_facet = $new_facet_obj;
```
Comment (misterbisson) [inline plugin/class-facets.php]: (I've yet to read all the code, so this is a note to myself)

What happens to URLs like http://search.gigaom.com/search/apple/?company=ibm ?

The result _should_ be http://search.gigaom.com/company/apple+ibm/ .

Comment (wluo) [inline plugin/class-facets.php]: good question! since we merge the newly converted facet with existing ones, this will become a search for both apple and ibm company facets: http://search.local.gostage.it/company/ibm+apple/

File: plugin/class-facets.php (hunk 1 above)
Comment (wluo) [inline plugin/class-facets.php]: sure. i just took the config name from the original ticket...

File: plugin/class-facets.php [hunk 3]
```diff
@@ -558,6 +576,91 @@ public function build_labels( $sing , $plur , $singcap = '' , $plurcap = '' )
...
+
+	/**
+	 * Check if we can redirect a keyword search to a faceted term search.
+	 * If so we'll set up our selected facets.
+	 *
+	 * @return mixed an array with two elements of [ new_facet_name, new_facet]
+	 *  if the current search is a keyword search and can be converted to a
+	 *  faceted term search, or FALSE if not.
+	 */
+	public function can_redirect_keyword_search()
```
Comment (misterbisson) [inline plugin/class-facets.php]: Should this live in `class-facet-searchword.php`? Does this method even need to be parsed and compiled if that class isn't loaded?

File: plugin/class-facet-searchword.php [hunk 4]
```diff
@@ -90,4 +90,40 @@ function permalink( $terms )
...
+
+		// get all taxonomy terms that match $search_term, as term objects
+		global $wpdb;
+
+		$terms = $wpdb->get_results(
+			$wpdb->prepare(
+				'SELECT t.term_id, t.name, t.slug, t.term_group,
+					tt.term_taxonomy_id, tt.taxonomy, tt.description,
+					tt.parent, tt.count
+				FROM ' . $wpdb->terms . ' t JOIN ' . $wpdb->term_taxonomy . ' tt
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: `JOIN`s should always be on a new line

File: plugin/class-facets.php (hunk 3 above)
Comment (wluo) [inline plugin/class-facets.php]: excellent point!

File: plugin/class-facet-searchword.php [hunk 5]
```diff
@@ -90,4 +90,40 @@ function permalink( $terms )
...
 		return get_search_link( $terms );
 	}
-}
+
+	/**
+	 * @param string $search_term the keyword search term slug
+	 * @return object authoritative term objects from $facets, or NULL if
+	 *  there're no authoritative terms in $facets
+	 */
+	public function get_authoritative_terms( $search_term )
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: This isn't getting the "authoritative" term, it's looking up terms. That the scriblio-authority plugin hooks to the filter applied below is incidental to the name and purpose.

Perhaps this name is misleading?

File: plugin/class-facets.php [hunk 6]
```diff
@@ -558,6 +576,91 @@ public function build_labels( $sing , $plur , $singcap = '' , $plurcap = '' )
...
+		{
+			return FALSE; // not a keyword search
+		}
+
+		// there should only be one term since don't break up the search
+		// terms in the search text box
+		$search_term = sanitize_title_with_dashes( array_keys( $this->selected_facets->searchword )[0] );
+
+		// check if we have cached facets for $search_term
+		$facets = wp_cache_get( $search_term, 'scriblio-facet' );
```
Comment (misterbisson) [inline plugin/class-facets.php]: `scriblio-facet` sounds like a cache group used for something else.

File: plugin/class-facets.php [hunk 7]
```diff
@@ -558,6 +576,91 @@ public function build_labels( $sing , $plur , $singcap = '' , $plurcap = '' )
...
+
+		if ( FALSE === $facets )
+		{
+			// do we have authority records for any of the search terms?
```
Comment (misterbisson) [inline plugin/class-facets.php]: Again, misleading use of "authority"

File: plugin/class-facets.php [hunk 8]
```diff
@@ -558,6 +576,91 @@ public function build_labels( $sing , $plur , $singcap = '' , $plurcap = '' )
...
+			$terms = $this->facets->searchword->get_authoritative_terms( $search_term );
+			if ( empty( $terms ) )
+			{
+				// cache negative results too
+				wp_cache_set( $search_term, array(), 'scriblio-facet', scriblio()->options['converted_facet_cache_ttl'] );
+				return FALSE;
+			}
+
+			// convert the authority terms back to facets
```
Comment (misterbisson) [inline plugin/class-facets.php]: Authority!

File: plugin/class-facets.php [hunk 9]
```diff
@@ -558,6 +576,91 @@ public function build_labels( $sing , $plur , $singcap = '' , $plurcap = '' )
...
+
+			// cache the results, even if they're empty
+			wp_cache_set( $search_term, $facets, 'scriblio-facet', scriblio()->options['converted_facet_cache_ttl'] );
+		}//END if
+		elseif ( empty( $facets ) )
+		{
+			return FALSE; // cache result is empty too
+		}
+
+		// pick out the facet with the highest priority (lowest number)
```
Comment (misterbisson) [inline plugin/class-facets.php]: Seems like this should move to a convenience method somewhere, perhaps in `class-facets.php`

File: plugin/class-scriblio.php [hunk 10]
```diff
@@ -16,6 +16,7 @@ class Scriblio
 			'widgets' => TRUE,
 		),
 		'register_default_facets' => TRUE,
+		'converted_facet_cache_ttl' => 604807, // ~7 days
```
Comment (misterbisson) [inline plugin/class-scriblio.php]: Feels like a lot of different terms are used for the same/similar thing throughout.

Here are some of the names I've seen:
- `redirect_searchword_to_taxonomy` (was `redirect_keywords_to_tax_terms`) from https://github.com/misterbisson/scriblio/pull/49#discussion_r16673861
- `scriblio-facet` in https://github.com/misterbisson/scriblio/pull/49#discussion_r16674941
- `get_authoritative_terms` from https://github.com/misterbisson/scriblio/pull/49#discussion_r16674887
- `can_redirect_keyword_search` from https://github.com/misterbisson/scriblio/pull/49#discussion_r16674664

File: plugin/class-scriblio.php [hunk 11]
```diff
@@ -157,6 +158,8 @@ public function get_default_facets()
 		// register public taxonomies as facets
 		foreach ( (array) get_taxonomies( array( 'public' => TRUE ) ) as $taxonomy )
 		{
+			$priority = isset( $this->options['default_facet_priorities'][ $taxonomy ] ) ? $this->options['default_facet_priorities'][ $taxonomy ] : 5;
```
Comment (misterbisson) [inline plugin/class-scriblio.php]: Pinning this to come back to. The past process was to register the facets using outside code and suppress Scriblio's registration of the facets.

File: plugin/class-scriblio.php (hunk 10 above)
Comment (misterbisson) [inline plugin/class-scriblio.php]: Perhaps...
- In `class-facets.php` it should call `$this->facets->searchword->to_taxonomy()`
- The cache group maybe should be `scriblio-searchword-to-taxonomy`

File: plugin/class-facets.php (hunk 2 above)
Comment (wluo) [inline plugin/class-facets.php]: does the order of terms matter? in your example, we'll actually end up with an url that looks like http://search.gigaom.com/company/ibm+apple/, because we added a new company facet "apple" to the existing company facet "ibm". Alternatively we could alphabetize the list of facet terms but I don't believe we're doing that now.

File: plugin/class-facet-searchword.php [hunk 12]
```diff
@@ -90,4 +90,134 @@ function permalink( $terms )
...
+
+	/**
+	 * Convert a keyword search term to a taxonomy (embodied in a facet)
+	 *
+	 * @return mixed an array with two elements of [ new_facet_name, new_facet]
+	 *  if the keyword search can be converted to a facet, or FALSE if not.
+	 */
+	public function to_taxonomy()
+	{
+		// there should only be one term since don't break up the search
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: Grammar?

File: plugin/class-facet-searchword.php [hunk 13]
```diff
@@ -90,4 +90,134 @@ function permalink( $terms )
...
+		// terms in the search text box
+		$search_slug = sanitize_title_with_dashes( array_keys( $this->facets->selected_facets->searchword )[0] );
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php:106]: This is sneaky smooth.

File: plugin/class-facet-searchword.php [hunk 14]
```diff
@@ -90,4 +90,134 @@ function permalink( $terms )
...
+		}
+		elseif ( FALSE === $facets )
+		{
+			// get all terms with slug $search_slug
+			$terms = $this->get_taxonomy_terms( $search_slug );
+
+			if ( empty( $terms ) )
+			{
+				// cache negative results too
+				wp_cache_set( $search_slug, array(), 'scriblio-searchword-to-taxonomy', scriblio()->options['converted_facet_cache_ttl'] );
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: The cache group is begging to be a classvar, no?

File: plugin/class-facet-searchword.php [hunk 15]
```diff
@@ -90,4 +90,134 @@ function permalink( $terms )
...
+			}
+
+			// convert the terms back to facets
+			$facets = scriblio()->get_terms_as_facets( $terms );
+
+			// cache the results, even if they're empty
+			wp_cache_set( $search_slug, $facets, 'scriblio-searchword-to-taxonomy', scriblio()->options['converted_facet_cache_ttl'] );
+		}//END if
+
+		// get facet/taxonmy name with the term with the highest count
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: spelling

File: plugin/class-facet-searchword.php [hunk 16]
```diff
@@ -90,4 +90,134 @@ function permalink( $terms )
...
+		if ( ! $new_facet_name = $this->get_most_popular_facet( $facets ) )
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: Running this _after_ `scriblio()->get_terms_as_facets( $terms )` requires looping in `get_most_popular_facet()` that could be avoided if the order were reversed maybe?

File: plugin/class-facet-searchword.php [hunk 17]
```diff
@@ -90,4 +90,134 @@ function permalink( $terms )
...
+	 * @param array $facets a list of facets to scan through
+	 * @return mixed name of the facet with the most-used term, or NULL
+	 *  if there is no term to work with.
+	 */
+	public function get_most_popular_facet( $facets )
+	{
+		$winner = NULL;
+		$max_count = -1;
+
+		foreach ( $facets as $facet_name => $terms )
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: Seeing all the foreaches hurts. See question in https://github.com/misterbisson/scriblio/pull/49#discussion_r16694839 .

Reason: I'd love to just do a `usort()` here and avoid the nested loops.

File: plugin/class-facets.php [hunk 18]
```diff
@@ -119,8 +117,29 @@ public function parse_query( $query )
...
 
+		// detect if a keyword search could be converted to a facet search
+		if (
+			isset( scriblio()->options['redirect_searchword_to_taxonomy'] ) &&
+			scriblio()->options['redirect_searchword_to_taxonomy'] &&
+			isset( $this->facets->searchword ) &&
+			isset( $this->selected_facets->searchword )
+		)
+		{
+			if ( list( $new_facet, $new_facet_obj ) = $this->facets->searchword->to_taxonomy() )
```
Comment (misterbisson) [inline plugin/class-facets.php]: I'm trying to decide if the searchword should be passed in here or not.

This clearly works, but not passing a value sometimes leads to methods that are hard to follow.

File: plugin/class-facet-searchword.php (hunk 14 above)
Comment (wluo) [inline plugin/class-facet-searchword.php]: that would've been my first choice but we don't seem to do this in many places.

File: plugin/class-facets.php (hunk 18 above)
Comment (wluo) [inline plugin/class-facets.php]: the tradeoff is that if we explicitly pass in the parameters (which would be `$this->facets->selected_facets` i think), then code from other classes can call this function too.

File: plugin/class-facet-searchword.php [hunk 19]
```diff
@@ -90,4 +92,131 @@ function permalink( $terms )
...
+				// cache negative results too
+				wp_cache_set( $search_slug, array(), $this->cache_group, scriblio()->options[ $this->cache_ttl_config_key ] );
+				return FALSE;
+			}
+
+			// iterate over $terms, which are now in descending count order,
+			// until we get a facet.
+			foreach ( $terms as $term )
+			{
+				$facets = scriblio()->get_terms_as_facets( array( $term ) );
```
Comment (wluo) [inline plugin/class-facet-searchword.php:137]: this is necessary because once the terms are converted to facets, they lose their ordering. (a facet contains an associative array of facet names to another associative array of term names => term objects.)

File: plugin/class-scriblio.php (hunk 10 above)
Comment (misterbisson) [inline plugin/class-scriblio.php]: This one doesn't match the config file: https://github.com/GigaOM/gigaom/pull/5114

File: plugin/class-facets.php (hunk 2 above)
Comment (misterbisson) [inline plugin/class-facets.php]: Yep, that looks right.

File: plugin/class-facets.php (hunk 18 above)
Comment (misterbisson) [inline plugin/class-facets.php]: > then code from other classes can call this function too

Yes, that's another argument for it.

Still, not sure I want to take this apart just to make that change.

File: plugin/class-facet-searchword.php (hunk 19 above)
Comment (misterbisson) [inline plugin/class-facet-searchword.php:137]: But there's a conditional break right below here, so it's all good.

Comment (wluo) [inline plugin/class-facet-searchword.php:137]: Yup we break as soon as we get a facet so we know it's the facet with the highest term count, since we're going down the list of terms sorted by count in descending order.

File: plugin/class-facets.php (hunk 18 above)
Comment (misterbisson) [inline plugin/class-facets.php]: Your thoughts? What would you prefer if you re-discover this code in six or nine months?

Comment (wluo) [inline plugin/class-facets.php]: I'm ok leaving it the way it is, because this call makes sense to me

`$this->facets->searchword->to_taxonomy()`

to me this means get the taxonomy for the current searchword facet.

File: plugin/class-facet-searchword.php [hunk 20]
```diff
@@ -2,10 +2,12 @@
 
 class Facet_Searchword implements Facet
 {
-
 	var $query_var = 's';
 	var $exclude_from_widget = TRUE;
 
+	private $cache_ttl_config_key = 'searchword_to_taxonomy_cache_ttl';
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: This would make sense if it was 

``` php
private $cache_ttl = 604807; // ~7 days
```

See also https://github.com/GigaOM/gigaom/pull/5114#discussion_r16811761

File: plugin/class-facet-searchword.php [hunk 21]
```diff
@@ -90,4 +92,131 @@ function permalink( $terms )
...
+			return FALSE; // empty cached result
+		}
+
+		// not found in our cache
+		if ( FALSE === $facets )
+		{
+			// get all terms with slug $search_slug
+			$terms = $this->get_taxonomy_terms( $search_slug );
+
+			// sort the terms by count since we passed them through the
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: I feel persnickety in saying the following, but the comment maybe misses the point: that the counts might have changed.

File: plugin/class-facet-searchword.php [hunk 22]
```diff
@@ -90,4 +91,131 @@ function permalink( $terms )
...
+			// its keys by a variable
+			$new_facet = (object) $this->facets->selected_facets->$new_facet_name;
+		}
+		else
+		{
+			$new_facet = new stdClass;
+		}
+
+		// copy over the query terms to the new facet
+		foreach ( $facets->$new_facet_name as $key => $val )
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: Can we do an array merge instead of foreach? This [appears to be an array](https://github.com/wluo/scriblio/blob/github-3995/plugin/class-facet-taxonomy.php#L50).

File: plugin/class-facet-searchword.php [hunk 23]
```diff
@@ -90,4 +91,131 @@ function permalink( $terms )
...
+		// merge $new_facet with the existing, selected facets
+		if ( isset( $this->facets->selected_facets->$new_facet_name ) )
+		{
+			// we need to cast the facet to an object so we can reference
+			// its keys by a variable
+			$new_facet = (object) $this->facets->selected_facets->$new_facet_name;
+		}
+		else
+		{
+			$new_facet = new stdClass;
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: You can do this, right?

``` php
$new_facet = (object) $facets->$new_facet_name;
```

File: plugin/class-facet-searchword.php [hunk 24]
```diff
@@ -90,4 +91,131 @@ function permalink( $terms )
...
+		}
+
+		$new_facet_name = array_keys( (array) $facets )[0];
+
+		// merge $new_facet with the existing, selected facets
+		if ( isset( $this->facets->selected_facets->$new_facet_name ) )
+		{
+			// we need to cast the facet to an object so we can reference
+			// its keys by a variable
+			$new_facet = (object) $this->facets->selected_facets->$new_facet_name;
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: What's the context for `$this`?

File: plugin/class-facet-searchword.php [hunk 25]
```diff
@@ -90,4 +91,131 @@ function permalink( $terms )
...
+			$new_facet = new stdClass;
+		}
+
+		// copy over the query terms to the new facet
+		foreach ( $facets->$new_facet_name as $key => $val )
+		{
+			$new_facet->$key = $val;
+		}//END foreach
+
+		return array( $new_facet_name, (array) $new_facet );
```
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: Return this, or just set `facets()->selected_facets` and return a boolean, given https://github.com/misterbisson/scriblio/pull/49#discussion_r16810114?

File: plugin/class-facet-searchword.php (hunk 24 above)
Comment (misterbisson) [inline plugin/class-facet-searchword.php]: An [array merge](https://github.com/misterbisson/scriblio/pull/49#discussion_r16812216) can be done here, no?

File: plugin/class-facets.php (hunk 18 above)
Comment (misterbisson) [inline plugin/class-facets.php]: Good point. So what about https://github.com/misterbisson/scriblio/pull/49#discussion_r16812361 ?

File: plugin/class-facet-searchword.php (hunk 23 above)
Comment (wluo) [inline plugin/class-facet-searchword.php]: you're right about this!

File: plugin/class-facet-searchword.php (hunk 22 above)
Comment (wluo) [inline plugin/class-facet-searchword.php]: we can indeed, and it not only gets rid of the `foreach` but also all those casts.

File: plugin/class-facet-searchword.php (hunk 24 above)
Comment (wluo) [inline plugin/class-facet-searchword.php]: `$this->facets` is the `Facets` object that contains this `Facet_Searchword` instance. it's passed in and set [in `Facet_Searchword`'s constructor](https://github.com/misterbisson/scriblio/blob/7c843df82ccca560c1a8fa641ee7e809d5f959b8/plugin/class-facet-searchword.php#L13).

**Comment by misterbisson** [comment]
:let's try it:
